README.md
![tn](https://github.com/k1350/ogpgen/blob/main/example/outsample.jpg)
## インストールとビルド
必要環境

- Git
- Go (version 1.19.3 で動作確認済)

```
git clone https://github.com/k1350/ogpgen.git
cd ogpgen
go build -o ogpgen ./cmd/ogpgen
```

## 使い方
最低限の引数指定時
```
./ogpgen -text="OGP 画像生成ツール" -bpath="/your_template/template.jpg" -fpath="/your_font/03SmartFontUI.ttf"
```

引数全指定時
```
./ogpgen -text="OGP 画像生成ツール" -bpath="/your_template/template.jpg" -fpath="/your_font/03SmartFontUI.ttf" -fsize=90 -fcolor="#A2Ad0580" -tmargin=75 -smargin=90 -lspace=10 -o="/out_dir/output.jpg"
```

### 引数の説明
- text: 【必須】描画したい文字列。
- bpath: 【必須】背景に使用する画像ファイルのパス。JPEG でなければならない。
- fpath: 【必須】使用するフォントファイルのパス。TrueType フォントでなければならない。
- fsize: フォントサイズ。0 より大きい数でなければならない。小数点以下も指定できる。未指定の場合は 100.0 になる。
- fcolor: フォント色を6桁または8桁（透明度指定あり）のカラーコードで指定する。未指定の場合は #000000 になる。
- tmargin: 上部マージン。文字が描画される位置が、正の値を指定すると下側に、負の値を指定すると上側にずれる。未指定の場合は 0 になる。
- smargin: 左右マージン。大きい値を指定するほど文字が中央に寄る。未指定の場合は 0 になる。
- lspace: 行間。正の値を指定すると行間が広がり、負の値を指定すると詰まる。未指定の場合は 0 になる。
- o: 出力ファイル名。未指定の場合は out.jpg になる。

## 開発
### テスト
このファイルがあるディレクトリの一つ上の階層に `ogpgen_files` というディレクトリを作り、下記の構成でファイルを置いておかないと一部テストが動作しない。

```
ogpgen_files
|
|-fonts
| |-03SmartFontUI.ttf
|
|-templates
  |-template.jpg
```

03SmartFontUI.ttf は TrueType フォントなら何でもいい。

template.jpg は OGP 画像のベースになる適当な jpg ファイル。


CHANGELOG.md
# CHANGELOG

## 0.0.4

- 依存関係の更新

## 0.0.3

- 内部で文字色の指定に color.NRGBA を使用するよう修正

## 0.0.2

- エラー発生時のログ書き出し内容を詳細にする修正

## 0.0.1

first release


internal/color/color_test.go
package color

import (
	"image/color"
	"reflect"
	"testing"
)

type tcase struct {
	in   string
	want color.NRGBA
}

func TestConvertHexToNRGBA(t *testing.T) {
	cases := []tcase{
		{
			in: "#FF0000",
			want: color.NRGBA{
				R: uint8(255),
				G: uint8(0),
				B: uint8(0),
				A: uint8(255),
			},
		},
		{
			in: "#00FF00",
			want: color.NRGBA{
				R: uint8(0),
				G: uint8(255),
				B: uint8(0),
				A: uint8(255),
			},
		},
		{
			in: "#0000FF",
			want: color.NRGBA{
				R: uint8(0),
				G: uint8(0),
				B: uint8(255),
				A: uint8(255),
			},
		},
		{
			in: "#FEC0AD80",
			want: color.NRGBA{
				R: uint8(254),
				G: uint8(192),
				B: uint8(173),
				A: uint8(128),
			},
		},
	}

	for _, c := range cases {
		got, _ := ConvertHexToNRGBA(c.in)
		assertColor(t, got, c.want)
	}
}

func assertColor(t *testing.T, c color.NRGBA, wc color.NRGBA) {
	var got [4]uint32
	got[0], got[1], got[2], got[3] = c.RGBA()
	var want [4]uint32
	want[0], want[1], want[2], want[3] = wc.RGBA()
	if !reflect.DeepEqual(got, want) {
		t.Errorf("got:\n%v\n\nwant:\n%v", got, want)
	}
}


internal/ogpgen/ogpgen_test.go
package ogpgen

import (
	"bufio"
	"image/jpeg"
	"log"
	"os"
	"testing"
)

func TestDraw(t *testing.T) {
	d, err := NewOgpImage(
		Option{
			FontPath:            "../../../ogpgen_files/fonts/03SmartFontUI.ttf",
			FontSize:            90.0,
			BackgroundImagePath: "../../../ogpgen_files/templates/template.jpg",
			TopMargin:           50,
			LineSpace:           10,
		},
	)
	if err != nil {
		log.Fatal(err)
	}

	img, err := d.Draw("あいAB!102")
	if err != nil {
		log.Fatal(err)
	}

	fn := "out.jpg"
	newFile, err := os.Create(fn)
	if err != nil {
		log.Fatal(err)
	}
	defer newFile.Close()
	b := bufio.NewWriter(newFile)
	if err = jpeg.Encode(b, img, &jpeg.Options{Quality: 100}); err != nil {
		log.Fatal(err)
	}
	if err := os.Remove(fn); err != nil {
		log.Fatal(err)
	}
}


internal/kinsoku/kinsoku_test.go
package kinsoku

import (
	"github.com/golang/freetype/truetype"
	"golang.org/x/image/font"
	"io/ioutil"
	"log"
	"reflect"
	"strings"
	"testing"
)

type tcase struct {
	in   string
	want []string
}

func TestExecKinsoku(t *testing.T) {
	ftBin, err := ioutil.ReadFile("../../../ogpgen_files/fonts/03SmartFontUI.ttf")
	if err != nil {
		log.Fatal(err)
	}
	ft, err := truetype.Parse(ftBin)
	if err != nil {
		log.Fatal(err)
	}
	face := truetype.NewFace(ft, &truetype.Options{Size: 1.0})

	width := 5

	t.Run("行頭禁則文字", func(t *testing.T) {
		cases := []tcase{
			{
				in:   "あいうえお、かきくけこ。",
				want: []string{"あいうえ", "お、かきく", "けこ。"},
			},
			{
				in:   "あお、〕:,゠ㇷ!.。かきくけこ",
				want: []string{"あ", "お、〕:,゠", "ㇷ!.。かき", "くけこ"},
			},
		}

		assertKinsoku(t, face, width, cases)
	})
	t.Run("行末禁則文字", func(t *testing.T) {
		cases := []tcase{
			{
				in:   "「はい」「いいえ」",
				want: []string{"「はい」", "「いいえ」"},
			},
			{
				in:   "あお（[｛〔“〘«《かきくけこ",
				want: []string{"あお（", "[｛〔“〘«", "《かきくけ", "こ"},
			},
		}

		assertKinsoku(t, face, width, cases)
	})
	t.Run("分離禁則文字", func(t *testing.T) {
		cases := []tcase{
			{
				in:   "人口は約125,36万人",
				want: []string{"人口は約", "125,36万", "人"},
			},
			{
				in:   "Lorem ipsum dolor sit amet",
				want: []string{"Lorem ", "ipsum ", "dolor sit ", "amet"},
			},
		}

		assertKinsoku(t, face, width, cases)
	})
}

func assertKinsoku(t *testing.T, face font.Face, width int, cases []tcase) {
	for _, c := range cases {
		got := ExecKinsoku(face, width, c.in)
		if !reflect.DeepEqual(got, c.want) {
			t.Errorf("got:\n%v\n\nwant:\n%v", strings.Join(got[:], "\n"), strings.Join(c.want[:], "\n"))
		}
	}
}


internal/color/color.go
// package color は 16 進数のカラーコードから color.NRGBA を生成するパッケージです。
package color

import (
	"image/color"
	"regexp"
	"strconv"

	e "github.com/k1350/ogpgen/internal/errors"
	"github.com/pkg/errors"
)

// rgbaFormat はアルファ値を含むカラーコードを判定するための正規表現です。
var rgbaFormat = regexp.MustCompile(`^#[0-9a-fA-F]{2}[0-9a-fA-F]{2}[0-9a-fA-F]{2}[0-9a-fA-F]{2}$`)

// rgbFormat はアルファ値を含まないカラーコードを判定するための正規表現です。
var rgbFormat = regexp.MustCompile(`^#[0-9a-fA-F]{2}[0-9a-fA-F]{2}[0-9a-fA-F]{2}$`)

// ConvertHexToNRGBA は 16 進数のカラーコードから color.NRGBA を生成して返します。
// hex はカラーコードです。#333 のような省略形は対応していません。
func ConvertHexToNRGBA(hex string) (color.NRGBA, error) {

	if ok := rgbaFormat.MatchString(hex); ok {
		return parseColor(hex, true)
	}
	if ok := rgbFormat.MatchString(hex); ok {
		return parseColor(hex, false)
	}
	return color.NRGBA{0, 0, 0, 0}, e.ErrorInvalidColorFormat
}

// parseColor は 16 進数のカラーコードから color.NRGBA を生成して返します。
// hex はカラーコードです。#333 のような省略形は対応していません。
// existAlpha はアルファ値を含むかどうかを指定します。true の場合はアルファ値を読み取り、false の場合はアルファ値は FF とします。
func parseColor(hex string, existAlpha bool) (c color.NRGBA, err error) {
	err = nil
	var a int64
	a = 255
	if existAlpha {
		a, err = strconv.ParseInt(hex[7:9], 16, 64)
		if err != nil {
			err = errors.Wrap(e.ErrorInvalidColorFormat, err.Error())
			return
		}
	}

	r, err := strconv.ParseInt(hex[1:3], 16, 64)
	if err != nil {
		err = errors.Wrap(e.ErrorInvalidColorFormat, err.Error())
		return c, err
	}
	g, err := strconv.ParseInt(hex[3:5], 16, 64)
	if err != nil {
		err = errors.Wrap(e.ErrorInvalidColorFormat, err.Error())
		return c, err
	}
	b, err := strconv.ParseInt(hex[5:7], 16, 64)
	if err != nil {
		err = errors.Wrap(e.ErrorInvalidColorFormat, err.Error())
		return c, err
	}

	c.R = uint8(r)
	c.G = uint8(g)
	c.B = uint8(b)
	c.A = uint8(a)
	return
}


internal/ogpgen/ogpgen.go
// package ogpgen は OGP 画像を生成するパッケージです。
package ogpgen

import (
	"image"
	"image/color"
	"image/draw"
	"image/jpeg"
	"io/ioutil"
	"os"

	"github.com/golang/freetype"
	"github.com/golang/freetype/truetype"
	c "github.com/k1350/ogpgen/internal/color"
	e "github.com/k1350/ogpgen/internal/errors"
	"github.com/k1350/ogpgen/internal/kinsoku"
	"github.com/pkg/errors"
	"golang.org/x/image/font"
	"golang.org/x/image/math/fixed"
)

// OgpImage は OGP 画像生成に必要なフィールド等を定義した構造体です。
type OgpImage struct {
	Font            *truetype.Font
	BackgroundImage image.Image
	FontSize        float64
	FontColor       color.NRGBA
	TopMargin       int
	SideMargin      int
	LineSpace       int
}

// Option は NewOgpImage の引数を定義したコンストラクタです。
// FontPath は OGP 画像に描画する文字列の TrueType フォントのパスを指定します。
// FontSize は OGP 画像に描画する文字列のフォントサイズを指定します。
// BackgroudImagePath は OGP 画像のベースになる jpeg 画像のパスを指定します。
// TopMargin は 文字列描画時の上部のマージンを微調整したい場合に指定します。
// SideMargin は 文字列描画時の左右のマージンを調整したい場合に指定します。
// LineSpace は 文字列描画時の行間を調整したい場合に指定します。
type Option struct {
	FontPath            string
	FontSize            float64
	FontColor           string
	BackgroundImagePath string
	TopMargin           int
	SideMargin          int
	LineSpace           int
}

// NewOgpImage は OgpImage のコンストラクタです。
func NewOgpImage(opt Option) (o *OgpImage, err error) {
	o = &OgpImage{}
	err = nil

	ftBin, err := ioutil.ReadFile(opt.FontPath)
	if err != nil {
		err = errors.Wrap(e.ErrorFontReadFailed, err.Error())
		return
	}
	ft, err := truetype.Parse(ftBin)
	if err != nil {
		err = errors.Wrap(e.ErrorFontParseFailed, err.Error())
		return
	}
	o.Font = ft

	if opt.FontSize <= 0.0 {
		err = e.ErrorFontSizeNegative
		return
	}
	o.FontSize = opt.FontSize

	if opt.FontColor == "" {
		opt.FontColor = "#000000"
	}
	c, err := c.ConvertHexToNRGBA(opt.FontColor)
	if err != nil {
		return
	}
	o.FontColor = c

	file, err := os.Open(opt.BackgroundImagePath)
	if err != nil {
		err = errors.Wrap(e.ErrorBackgroundImageReadFailed, err.Error())
		return
	}
	defer file.Close()
	img, err := jpeg.Decode(file)
	if err != nil {
		err = errors.Wrap(e.ErrorBackgroundImageDecodeFailed, err.Error())
		return
	}
	o.BackgroundImage = img

	o.TopMargin = opt.TopMargin
	o.SideMargin = opt.SideMargin
	o.LineSpace = opt.LineSpace

	return
}

// Draw は OGP 画像のベース画像に指定された文字列 text を描画します。
func (o *OgpImage) Draw(text string) (img *image.RGBA, err error) {
	rect := image.Rectangle{image.Pt(0, 0), o.BackgroundImage.Bounds().Size()}
	img = image.NewRGBA(rect)
	draw.Draw(img, img.Bounds(), o.BackgroundImage, image.ZP, draw.Src)

	c := freetype.NewContext()
	c.SetDPI(72)
	c.SetFont(o.Font)
	c.SetFontSize(o.FontSize)
	c.SetClip(img.Bounds())
	c.SetDst(img)
	c.SetSrc(image.NewUniform(o.FontColor))
	c.SetHinting(font.HintingNone)

	imageWidth := o.BackgroundImage.Bounds().Dx()
	imageHeight := o.BackgroundImage.Bounds().Dy()
	face := truetype.NewFace(o.Font, &truetype.Options{Size: o.FontSize})

	words := kinsoku.ExecKinsoku(face, (imageWidth - o.SideMargin*2), text)

	rows := len(words)
	for i, v := range words {
		textWidth := font.MeasureString(face, v).Ceil()
		dot := fixed.Point26_6{X: fixed.I((imageWidth - textWidth) / 2), Y: fixed.I((imageHeight-face.Metrics().Height.Ceil()*rows)/2 + (i+1)*(face.Metrics().Height.Ceil()+o.LineSpace) + o.TopMargin)}
		_, err = c.DrawString(v, dot)
		if err != nil {
			err = errors.Wrap(e.ErrorDrawFailed, err.Error())
			return
		}
	}

	return
}


internal/kinsoku/kinsoku.go
// Package kinsoku は禁則処理を行うパッケージです。
package kinsoku

import (
	"bytes"
	"regexp"

	"golang.org/x/image/font"
)

// invalidStart は行頭禁則文字を定義した map です。
var invalidStart = map[string]bool{
	",": true, ")": true, "]": true, "｝": true, "、": true, "〕": true, "〉": true, "》": true, "」": true, "』": true, "】": true, "〙": true, "〗": true, "〟": true, "’": true, "”": true, "｠": true, "»": true,
	"ゝ": true, "ゞ": true, "ー": true, "ァ": true, "ィ": true, "ゥ": true, "ェ": true, "ォ": true, "ッ": true, "ャ": true, "ュ": true, "ョ": true, "ヮ": true, "ヵ": true, "ヶ": true,
	"ぁ": true, "ぃ": true, "ぅ": true, "ぇ": true, "ぉ": true, "っ": true, "ゃ": true, "ゅ": true, "ょ": true, "ゎ": true, "ゕ": true, "ゖ": true, "ㇰ": true, "ㇱ": true, "ㇲ": true, "ㇳ": true, "ㇴ": true,
	"ㇵ": true, "ㇶ": true, "ㇷ": true, "ㇸ": true, "ㇹ": true, "ㇷ゚": true, "ㇺ": true, "ㇻ": true, "ㇼ": true, "ㇽ": true, "ㇾ": true, "ㇿ": true, "々": true, "〻": true,
	"‐": true, "゠": true, "–": true, "〜": true, "～": true,
	"?": true, "!": true, "‼": true, "⁇": true, "⁈": true, "⁉": true,
	"・": true, ":": true, ";": true, "/": true, "。": true, ".": true,
}

// invalidEnd は行末禁則文字を定義した map です。
var invalidEnd = map[string]bool{
	"(": true, "[": true, "｛": true, "〔": true, "〈": true, "《": true, "「": true, "『": true,
	"【": true, "〘": true, "〖": true, "〝": true, "‘": true, "“": true, "｟": true, "«": true,
}

// 分離禁則の単語を検出するための正規表現です。
var notSplit = regexp.MustCompile(`^([a-zA-Z]{2}|,[0-9]|[0-9],|[0-9]{2}|——|……|‥‥|〳〳|〴〴|〵〵)$`)

// ExecKinsoku は禁則処理を行いながら半角文字相当の文字幅 width を超えないように文字列 text を改行します。
// 改行した結果は配列で返します。
func ExecKinsoku(face font.Face, width int, text string) (words []string) {

	var tmpbuf bytes.Buffer
	for _, r := range text {

		tmpbuf.WriteString(string(r))
		for {
			if font.MeasureString(face, tmpbuf.String()).Ceil() > width {
				tmpr := []rune(tmpbuf.String())
				splitidx := determineSplitIndex(tmpr, len(tmpr)-1)
				words = append(words, string(tmpr[:splitidx]))
				tmpbuf.Reset()
				tmpbuf.WriteString(string(tmpr[splitidx:]))
			}
			// 改行位置が確定後、残った文字列が文字幅以下なら次の文字へ進む
			if font.MeasureString(face, tmpbuf.String()).Ceil() <= width {
				break
			}
		}
	}
	if tmpbuf.Len() > 0 {
		words = append(words, tmpbuf.String())
	}
	return words
}

// determineSplitIndex は、r に対して禁則処理を考慮して改行位置を決定します。
func determineSplitIndex(r []rune, defaultidx int) (idx int) {
	idx = defaultidx
	for {

		// idx の位置で改行可能なら改行位置を確定する
		if ok := canLineBreak(r, idx); ok {
			break
		}

		// 改行位置を更に前にずらす
		idx--

		// 前の行全部を次の行に回しても改行できない場合は諦めて最初の改行位置で改行する
		if idx < 1 {
			idx = defaultidx
			break
		}
	}
	return
}

// canLineBreak は、r を idx の位置で改行可能かどうかを返します。
func canLineBreak(r []rune, idx int) bool {
	nextr := r[idx:]
	// 次の行の行頭が行頭禁則文字なら改行不可
	if _, ok := invalidStart[string(nextr[:1])]; ok {
		return false
	}

	prevr := r[:idx]
	prevend := len(prevr) - 1
	// 前の行の行末が行末禁則文字なら改行不可
	if _, ok := invalidEnd[string(prevr[prevend:])]; ok {
		return false
	}

	// 分離禁則チェック
	s := string(prevr[prevend:]) + string(nextr[:1])
	if ok := notSplit.MatchString(s); ok {
		return false
	}

	return true
}


cmd/ogpgen/main.go
package main

import (
	"bufio"
	"flag"
	"image/jpeg"
	"log"
	"os"

	e "github.com/k1350/ogpgen/internal/errors"
	"github.com/k1350/ogpgen/internal/ogpgen"
	"github.com/pkg/errors"
)

func main() {
	text := flag.String("text", "", "Output Text")
	fpath := flag.String("fpath", "", "Font Path")
	fsize := flag.Float64("fsize", 100.0, "Font Size")
	fcolor := flag.String("fcolor", "#000000", "Font Color Code")
	bpath := flag.String("bpath", "", "Background Image Path")
	tmargin := flag.Int("tmargin", 0, "Top Margin")
	smargin := flag.Int("smargin", 0, "Side Margin")
	lspace := flag.Int("lspace", 0, "Line Space")
	out := flag.String("o", "out.jpg", "Output File Path")
	flag.Parse()

	d, err := ogpgen.NewOgpImage(
		ogpgen.Option{
			FontPath:            *fpath,
			FontSize:            *fsize,
			FontColor:           *fcolor,
			BackgroundImagePath: *bpath,
			TopMargin:           *tmargin,
			SideMargin:          *smargin,
			LineSpace:           *lspace,
		},
	)
	if err != nil {
		log.Fatalf("%+v", err)
	}

	img, err := d.Draw(*text)
	if err != nil {
		log.Fatalf("%+v", err)
	}

	newFile, err := os.Create(*out)
	if err != nil {
		err = errors.Wrap(e.ErrorOutputFileCreateFailed, err.Error())
		log.Fatalf("%+v", err)
	}
	defer newFile.Close()

	b := bufio.NewWriter(newFile)
	err = jpeg.Encode(b, img, &jpeg.Options{Quality: 100})
	if err != nil {
		err = errors.Wrap(e.ErrorOutputFailed, err.Error())
		log.Fatalf("%+v", err)
	}
}
